VACMS-11505: add Last Saved by an Editor field to all  content types

edmund-dunn opened the pull request:
## Description

Closes #11505.

## Testing done
Tested locally. 

## Screenshots
![CleanShot 2022-11-15 at 08 10 04](https://user-images.githubusercontent.com/109987005/201969261-e6ca299a-a04d-4628-b6ad-a071cc65762c.png)
![CleanShot 2022-11-15 at 08 07 53](https://user-images.githubusercontent.com/109987005/201968812-c8b0a3f0-4c09-4d41-b182-4b03753bbf49.png)

## QA steps

Select a piece of content, scroll to the bottom and change the moderation state to Published and click save. In the node view, you can see the field at the bottom of the right sidebar with a date and time. If you go back into the edit form, you will be able to see the information in the top section of the right sidebar. 

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [ ] `Platform CMS Team`
- [ ] `Sitewide program`
- [x] `⭐️ Sitewide CMS`
- [ ] `⭐️ Public websites`
- [ ] `⭐️ Facilities`
- [ ] `⭐️ User support`


### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping the UX writer so they are ready to publish after deployment
- [ ] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping the UX writer to write and queue content
- [ ] No announcement is needed for this code change.
  - [ ] Merge & carry on unburdened by announcements

swirtSJW commented on config/sync/core.entity_form_display.node.health_care_local_facility.default.yml:
```diff
@@ -55,7 +56,6 @@ third_party_settings:
     group_editorial_workflow:
       children:
         - moderation_state
-        - revision_log
```
This is AWOL.    Needs to be added back by hand.

swirtSJW replied:
This has to be put back  or the test will never pass.

ndouglas commented on docroot/modules/custom/va_gov_backend/src/EventSubscriber/LastEditorSaveEventSubscriber.php:
```diff
@@ -0,0 +1,58 @@
+   *   The form state.
+   */
+  public function lastSavedByEditorSetTimestamp(array $form, FormStateInterface $form_state) {
+    $node = $form_state->getFormObject()->getEntity();
```
Just needs a type hint to get rid of the PHPStan gripe.
![Screen Shot 2022-11-10 at 7 16 39 AM](https://user-images.githubusercontent.com/1318579/201089349-e480b8c8-77c4-418c-b53c-581af3878695.png)

ndouglas commented on docroot/modules/custom/va_gov_backend/src/EventSubscriber/LastEditorSaveEventSubscriber.php:
```diff
@@ -0,0 +1,58 @@
+   */
+  public function formAlter(FormAlterEvent $event): void {
+    $form = &$event->getForm();
+    $form_state = $event->getFormState();
```
Do you need `$form_state` for anything?

ndouglas reviewed (changes requested):
Looks great, just a couple quibbles from the checks.

EWashb commented:
All functionality looks/acts as expected in my QA

EWashb reviewed (dismissed):
all looks/functions as expected in the QA steps.

swirtSJW commented on docroot/modules/custom/va_gov_backend/va_gov_backend.module:
```diff
@@ -1003,6 +1002,8 @@ function va_gov_backend_preprocess_page(&$variables) {
       'vet_center_mobile_vet_center',
       'vet_center_outstation',
     ];
+    $last_saved_by_an_editor = $node->get('field_last_saved_by_an_editor')->value;
```
When the field is empty, it is just picking up the current date time.  This needs to be defensive about picking up a non value.
![image](https://user-images.githubusercontent.com/5752113/201374418-662294c8-2494-4d87-af5b-3de054598d59.png)

![image](https://user-images.githubusercontent.com/5752113/201374466-b3b8cffa-30d5-46fd-a16a-a6b971c0fac2.png)

edmund-dunn replied:
@swirtSJW explain further

edmund-dunn replied:
@swirtSJW ok, now I got it. This has been addressed. I added a ternary operator.

swirtSJW reviewed (changes requested):
This is not working as expected.  When empty, it is displaying current time and date 

Other issues, summary from individual comments:

1. We have a bunch of unexprected changes coming through in the config, some of which were not there the on your first config exports.

   - lazy loading - I think that one is related to a dependabot bump of blazy, so it is unexpected but I thnk innocuouse.   - neds a QA step to double check one of the affected content types just to be sure.
   - field_administration going missing from hiddden.. that one seeems odd and should definitely be double checked for the impact.  
   - loss of meta title from groupping,  should be doublle checked 
    - There is a very wierd addition about time limit.   This one definetly needs checking out 
 2. looose check for string node is casting too wide a net.  See suggested code.
  3. Auto formatting of behat spec tests is making it look like every line changed.  Set your IDE to not format .feature files.  We want the raw output from spec to persist so that we can see only changes to the line that changed.
  4. QA steps are outdated please adjust them to the changes made
  5. Add a QA step to cover checking the dates on nodes that have not been editied sincee this change.

edmund-dunn commented on docroot/modules/custom/va_gov_backend/va_gov_backend.module:
```diff
@@ -1003,6 +1002,8 @@ function va_gov_backend_preprocess_page(&$variables) {
       'vet_center_outstation',
     ];
+    $last_saved_by_an_editor = $node->get('field_last_saved_by_an_editor')->value;
+    $variables['page']['last_saved_by_an_editor'] = $last_saved_by_an_editor ? \Drupal::service('date.formatter')->format($last_saved_by_an_editor, 'custom', 'F j Y g:ia') : 'This needs to be reviewed by an editor.';
```
@ndouglas no, it is the format used in the recent changes sidebar.

ndouglas replied:
Should we formalize it?  Feels weird having it in code like a magic constant.

edmund-dunn commented on composer.json:
```diff
@@ -200,7 +200,7 @@
         "symfony/routing": "^4.0",
         "traviscarden/behat-table-comparison": "~0.2",
-        "va-gov/content-build": "^0.0.3136",
+        "va-gov/content-build": "^0.0.3138",
```
I merged in main.

laflannery reviewed (commented):
The "Updated By:" field seems a bit misleading.  I believe this is actually the author, the original author, that's what it is in Prod currently.  However it seems like the label was updated here but not the functionality.  For example, I updated a piece of content and the new human field has the correct date and time but it still says updated by CMS Migrator:
![image](https://user-images.githubusercontent.com/106678594/201980523-ebe61d50-ed48-4f17-a0d5-9b21dab82bb8.png)

swirtSJW commented:
This message seems a little misleading
![image](https://user-images.githubusercontent.com/5752113/202463251-1e734bad-a560-4c78-8c1e-b80bb33d03ae.png)

It was saved by an editor, just not since the field was added.  I realize some of this will go away when we bulk process the published and draft nodes to update that field, but until that happens, AND for the archived nodes that we are not going to touch, this message is not speaking the truth, which leads to confusion and cognition errors.   @BlakeOrgan @EWashb

EWashb commented:
> This message seems a little misleading ![image](https://user-images.githubusercontent.com/5752113/202463251-1e734bad-a560-4c78-8c1e-b80bb33d03ae.png)
> 
> It was saved by an editor, just not since the field was added. I realize some of this will go away when we bulk process the published and draft nodes to update that field, but until that happens, AND for the archived nodes that we are not going to touch, this message is not speaking the truth, which leads to confusion and cognition errors. @BlakeOrgan @EWashb
> 
> Suggestions:
> 
> * "Unknown"
> * "Uncertain"
> * "?"
> * "Not recorded"
> * (leave it empty)
> * (hide the display )

But the editor was a human editor in this example? I thought that user was a bot. Maybe we say something to signify that the last save was a bot.

edmund-dunn commented:
@EWashb in this case it because it hasn't been saved since we added the field.

EWashb commented:
That's exactly what I was thinking. Maybe we do a slight change in wording here, but possibly being pressed for time makes me wonder if we just hide it for the moment so that we can get an accurate report done before the sprint is over.

edmund-dunn commented:
I think just put "Unknown" since next sprint it will mostly go away with the script to update the field.

BlakeOrgan commented:
I think using "Unknown" is the best option. Once that's updated, I think that this would be good to approve from my end.

edmund-dunn commented:
> The "Updated By:" field seems a bit misleading. I believe this is actually the author, the original author, that's what it is in Prod currently. However it seems like the label was updated here but not the functionality. For example, I updated a piece of content and the new human field has the correct date and time but it still says updated by CMS Migrator: ![image](https://user-images.githubusercontent.com/106678594/201980523-ebe61d50-ed48-4f17-a0d5-9b21dab82bb8.png)

I changed this to be the latest revision author after conferring with @EWashb

laflannery reviewed (approved):
Looks good!

swirtSJW commented:
I am a bit concerned that this got merged without an passing approval from an engineer.
![image](https://user-images.githubusercontent.com/5752113/202539119-3cf7cd0c-b421-400c-abdc-f7e8987aeab0.png)

Especially when the last requested change were engineering related.
